Make `PowerManagingActor` more modular and generic

shsms opened the pull request:
This PR updates the `PowerManagingActor` to

- take type of components we are controlling. Currently only batteries
  are supported.
- Remove all other references to batteries from the
  `PowerManagingActor`, except in the interactions with the
  `PowerDistributingActor`.

The `PowerDistributingActor` needs to be modularized as well, before
the either of these actors can be used with other types of components.

All changes in this PR are internal and don't affect the external
interface of the `BatteryPool`.

shsms committed:
Rename `BatteryPool.PowerMetrics` -> `SystemBounds`

Because it an internal type that only holds system bounds.  It can be
used for other pools also, so it will be moved to a separate module in
a subsequent commit.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Move `battery_pool.SystemBounds` to `timeseries._base_types`

This allows the type to be used by the other component pools and
reduces the PowerManagingActor's dependence on the battery_pool
package.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Remove mentions of batteries from PowerManager's ReportRequest class

And replace them with the word `component`, so that it can be used as
a generic report request type for all pools.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Override `PowerManager.Report` with protocol class `BatteryPoolReport`

This allows the generic `Report` class to be typed in the BatteryPool as
`BatteryPoolReport` with battery-related documentation.

This commit also removed mentions of batteries from the
`PowerManager.Report` class docstrings.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Take bounds from the correct component pool in PowerManagingActor

This is in preparation for using the PowerManagingActor with different
component pools.

The power manager needs to get system bounds for the given set of
components from the right place, and only when necessary.  This
requires telling the power manager which pool to get bounds from, so
that it can do it dynamically.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Remove references to batteries from the PowerManagingActor

The only ones that remain are in the interactions with the
PowerDistributingActor, which will be made general purpose separately.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

llucax commented on src/frequenz/sdk/timeseries/_base_types.py:
```diff
@@ -150,3 +152,37 @@ class Bounds(Generic[_T]):
+    """
+
+
+class PoolType(enum.Enum):
+    """Enumeration of component pool types."""
+
+    BATTERY_POOL = "BATTERY_POOL"
+    EV_CHARGER_POOL = "EV_CHARGER_POOL"
+    PV_POOL = "PV_POOL"
+    CHP_POOL = "CHP_POOL"
```
Is there any way to make this using a class instead where we can use different subclasses to customize the behavior for different types? Usually using enums makes the code harder to read in maintain, as it gets littered with `if`s (or now `match`es probably), or is this another case where the type will need to be sent over the wire?

Sorry, I'm not very familiar with this code to evaluate it myself :-/

But even if the type is sent over the wire, we could do the implementation using these "trait" classes, so we just have to map enum -> class now, but then the code still is clean and without a lot of `if`s for having different logic for different types.

shsms replied:
It is for accessing the instance of the pool, in here: https://github.com/frequenz-floss/frequenz-sdk-python/blob/aebdd00b19f6b8fe38eca3a2604e04923f14cf70/src/frequenz/sdk/actor/_power_managing/_power_managing_actor.py#L133-L137

It would be an if statement at startup, but I think any alternative would be a lot more work,  and I think this is intuitive, and that `_add_bounds_tracker` function would be the only place such an if-statement is necessary, so it is quite isolated as well.

And there won't be edge cases because PowerManager instances are created internally,  so we can control and test how and where it is initialized.

llucax commented on src/frequenz/sdk/timeseries/_base_types.py:
```diff
@@ -150,3 +152,37 @@ class Bounds(Generic[_T]):
     """Upper bound."""
+
+
+@dataclass(frozen=True)
```
I would use `kw_only=True` here, as we have 2 attributes with `Bounds[Power]` type, that could be easily mixed up.

shsms replied:
https://github.com/frequenz-floss/frequenz-sdk-python/commit/ebb0e2d2

llucax commented on src/frequenz/sdk/timeseries/battery_pool/_result_types.py:
```diff
@@ -1,45 +1,90 @@
-    """Power bounds metrics."""
+# This class is used to expose the generic reports from the PowerManager with specific
+# documentation for the battery pool.
+class BatteryPoolReport(Protocol):
```
I didn't get how this works, why do we need a `Protocol` here? I'm concerned about starting to use protocols everywhere because then we lose the relationship between types, if `BateryPoolReport` inherits from `Report` is very clear what's going on, and we'll also get type/attributes mismatch errors in the class declaration, not when the class is used.

Also this seems a bit backwards, usually you declare a `Protocol` that is a general thing, and then the specific thing follows that protocol, so what seems more natural is to have `Report` as the protocol and `BatteryPoolReport` just a plan class, that follows that protocol.

I guess what you want to achieve here is to offer a "view" on some other type, right? If so it seems like a neat idea but IMHO is hard to reason about, so I would try to find a more common/pythonic way to achieve the same effect. What about:

```py
class Report(Protocol):
    # Interface only

class _Report(Report):
    # Concrete type with the implementation

class BatteryPoolReport(Report):
    # Interface with docs adapted to battery pools
```

Then for all the type hints for the generic object we use `Report` and when only when instantiating we use `_Report`.

shsms replied:
done in: https://github.com/frequenz-floss/frequenz-sdk-python/commit/88caff36

llucax reviewed (commented):
LGTM in general, but I have a few design concerns.

shsms committed:
Derive `Report` implementation and interface from a protocol class

This is supposed to establish a link between the classes, making it
easy to understand the code.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Make internal `SystemBounds` class `kw_only`

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

daniel-zullo-frequenz commented on tests/timeseries/_battery_pool/test_battery_pool_control_methods.py:
```diff
@@ -150,14 +151,9 @@ def _assert_report(
-        assert report._inclusion_bounds.lower == Power.from_watts(lower)
-        assert report._inclusion_bounds.upper == Power.from_watts(upper)
-        assert report._exclusion_bounds.lower == Power.from_watts(0.0)
-        assert report._exclusion_bounds.upper == Power.from_watts(0.0)
+        assert report.bounds is not None
+        assert report.bounds.lower == Power.from_watts(lower)
+        assert report.bounds.upper == Power.from_watts(upper)
         # pylint: enable=protected-access
```
disable/enable protected-access might be no longer needed

shsms replied:
done: https://github.com/frequenz-floss/frequenz-sdk-python/commit/91a3c825

shsms committed:
Remove unused pylint hints

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/_result_types.py:
```diff
@@ -1,45 +1,90 @@
-    """Power bounds metrics."""
+# This class is used to expose the generic reports from the PowerManager with specific
+# documentation for the battery pool.
+class BatteryPoolReport(Report):
```
IIUC this change should be mentioned in the release notes because this is the inner type that `power_status()` will return and the user needs to be aware of but I might be missing something here. Can you please check it out?

shsms replied:
true, done here: 20b45702

daniel-zullo-frequenz reviewed (commented):
I only have a question about a possible missing entry into the release notes. LGTM otherwise

daniel-zullo-frequenz reviewed (commented):
I only have a question about a possible missing entry into the release notes. LGTM otherwise

shsms committed:
Merge branch 'v1.x.x' into generic-power-manager

The RELEASE_NOTES.md file has diverged and needs to be synced from the
main branch, before it can be updated without conflicts.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Update RELEASE_NOTES.md

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>
